Move CLI actions and parsers to cli_utils

pdmnyberg opened the pull request:
This is an example of how you could setup the CLI utils to be shared with the web app.

pdmnyberg committed:
Move CLI actions and parsers to cli_utils

nimarafati commented on PLP_directRNA_design_package/plp_directrna_design/probedesign.py:
```diff
@@ -262,13 +262,7 @@ def parse_gtf(gtf_file, genes_str=None, identifier_type='gene_id'):
-    else:
-        genes_of_interest = None
-        raise InputValueError("No gene list provided. Processing all genes. $genes_of_interest", field="genes_of_interest", code="no_gene_list_provided")
+    print(f"Processing genes: {', '.join(genes_of_interest)}")
```
- There is no check for spliting the genes provided in a  comma separated format.
- Also no error handling?

nimarafati replied:
The idea here is to parse the string of genes (gene1, gene2,...) before passing it to the function in a form of array

pdmnyberg replied:
This is true. The function should atleast check that `genes_of_interest` is an iterable.

nimarafati commented on PLP_directRNA_design_package/plp_directrna_design/probedesign.py:
```diff
@@ -285,7 +279,7 @@ def parse_gtf(gtf_file, genes_str=None, identifier_type='gene_id'):
         raise InputValueError("No matching genes found in the GTF file. This can be due to either incomplete gtf file or errors in gene identifications.", field="genes_of_interest", code="no_matching_genes_found")
 
-    return gtf_df, genes_of_interest
+    return gtf_df
```
Here we need to return `genes_of_interest` since it will be used [here](https://github.com/Moldia/PLP_directRNA_design_V2/blob/e0de878ee8e6ab5eecefc317a2c0890cb088f7b1/codes/extract_features.py#L12)

pdmnyberg replied:
We resolved this discussion determining that `genes_of_interest` will be provided externally. Right?

nimarafati replied:
Yes, in the updated version it has been resolved so ignore my comment

nimarafati committed:
Update python version fot compatibility with cli_utils

nimarafati committed:
Import probedesign.py from plp_deisgn_package

nimarafati committed:
Import probedesign.py from cli_util

nimarafati committed:
Import adapterindex.py from cli_util

nimarafati committed:
Add master_parser

nimarafati committed:
Testing cli functions

nimarafati committed:
Incorporate updated cutadapt 2506

nimarafati committed:
Fix gene_str set/str type

nimarafati committed:
Check if isoforms.empty

Some features may not have more than one isoforms which means that the coverage cannot be calculated.   So by this check we skip the coverage calculation

nimarafati committed:
Update find_probes_in_targets and find_all

Updated these function to adapt new version of cutadapt implementation developed by marcel
At the moment it runs smoothly with 1 - 2 mismatches but it becomes exponentially slow by increasing the number of mismatches.
This will be improved under issue 39.

nimarafati committed:
Rename fasta_file to sequences_output in find_target function and parser

nimarafati committed:
Adjust parameters for find_target to read extracted regions/targets

pdmnyberg commented on codes/run_plp_directrna.py:
```diff
@@ -0,0 +1,91 @@
+from plp_directrna_design import cli_utils as cli
+import argparse
+
+def main():
```
I would like to convert this function to a reusable function as well. Because it is really this function that represents what we will be doing in the webapp right? As a function maybe called `cli.run_plp_directrna(...)`

nimarafati replied:
Sounds good!

pdmnyberg commented on codes/run_plp_directrna.py:
```diff
@@ -0,0 +1,91 @@
+
+    # # Step 4: Find targets
+    print("FINDING TARGETS")
+    if args.iupac_mismatches == "None":
```
I feel like we should make sure the parser parses this value properly rather than having this check.

nimarafati replied:
This was a quick fix so the user could disable the IUPAC check by passing "None" as a string. But the downstream function expects a Python None, so we need to convert the string "None" to the actual None value.

So if You think this can be checked in the parser please move it

pdmnyberg replied:
I noticed that we have a `master_parser` in the `cli_utils` already. I assume that that one is an older version which can be replaced with this one.

nimarafati replied:
Yes that's true! I had forgotten about that!

pdmnyberg commented on PLP_directRNA_design_package/plp_directrna_design/cli_utils.py:
```diff
@@ -0,0 +1,1378 @@
+def parse_genes(genes_str: str):
+    print(f"Parsing genes: {genes_str}")
+    selected_genes = set([g.strip().lower() for g in genes_str.split(",")])
+#    unknown_genes = selected_genes - full_gene_set
```
@nimarafati do you want me to remove the `gene_list` and be more lax with the requirements here? We should remove the commented code if it is not wanted.

nimarafati replied:
Sure! Please adjust as it will be more efficient.

pdmnyberg committed:
refactor: move run_plp_directrna to cli_utils

pdmnyberg committed:
fix: remove unused gene list from cli_utils

pdmnyberg commented on PLP_directRNA_design_package/plp_directrna_design/probedesign.py:
```diff
@@ -264,11 +264,15 @@ def parse_gtf(gtf_file, genes_str=None, identifier_type='gene_id'):
     if genes_str:
-            genes_of_interest = set([g.strip().lower() for g in genes_str.split(",")])
-            print(f"Processing genes: {', '.join(genes_of_interest)}")
+            if isinstance(genes_str, str):
```
@nimarafati The purpose of changing `gene_str` to be `genes_of_interest` instead was as follows:
- Move the parsing and error checking of the input string to an earlier stage
- Enable sharing the code for parsing of `gene_str`
- Decouple the parsing of `genes_of_interest` from `parse_gtf`

(Since I wasn't completely familiar with your code I opted to not replace `gene_str` initially and then I forgot about the details here.)

What I want to get is that if you think that it is important for function to be able to parse `gene_str` by themselves instead of taking a `list[str]` as input I think you should atleast make sure that all function that use `gene_str` share the same method of parsing it. For example by adding a `parse_gene_str` function. (Personally I think that using a `genes_of_interest: list[str]` would be the better choice but I assume I am missing some context here in order to make a complete judgment.)

Currently we have two different implementations of the parsing of `gene_str` in `parse_gtf` and `parse_gtf_with_gene_feature` and I would consider the cause of that to be unnecessary code duplication.

pdmnyberg committed:
refactor: clean up code for simplified use in web app adapter

pdmnyberg committed:
fix: fix regression in independent command line tools

nimarafati commented on PLP_directRNA_design_package/plp_directrna_design/probedesign.py:
```diff
@@ -113,10 +89,6 @@ def evaluate_ligation_junction(targets, iupac_mismatches=None, plp_length=30):
         targets.loc[idx, 'Ligation junction'] = ligation_status
 
         if iupac_mismatches is not None:
```
When I run the code, it seems `iupac_mismatches` is not properly set to `None`. I received this error :
```
File "/app/PLP_directRNA_design_package/plp_directrna_design/probedesign.py", line 94, in evaluate_ligation_junction
    raise InputValueError("The number of mismatches should be less than or equal to 2",
plp_directrna_design.probedesign.InputValueError: The number of mismatches should be less than or equal to 2```

pdmnyberg committed:
fix: use iupac parser for thr run_plp_directrna script

nimarafati committed:
Fix the reporting of probes with dinucleotide repetition

nimarafati committed:
Handling Tm=None

nimarafati committed:
Remove `default=None` for melting temp. params

nimarafati committed:
Merge pull request #41 from Moldia/experimental/add-cli-utils-Tm-None

Handling Tm=None

nimarafati committed:
Save off_target.csv

nimarafati committed:
Save off_targets

nimarafati committed:
Merge branch 'experimental/add-cli-utils' of github.com:moldia/PLP_directRNA_design_V2 into experimental/add-cli-utils

pdmnyberg commented on PLP_directRNA_design_package/plp_directrna_design/probedesign.py:
```diff
@@ -922,7 +922,7 @@ def find_targets(selected_features, sequences_output, reference_fasta, plp_lengt
     # Introduce IUPAC mismatches if specified and check ligation junctions. Or just check ligation junctions
     targets_df = evaluate_ligation_junction(targets_df, iupac_mismatches=iupac_mismatches, plp_length=plp_length)
-    off_target_info = None  # default when specificity is not checked
+    #off_target_info = None  # default when specificity is not checked
```
I think this change broke this function for the case when `check_specificity` is false. The variable still needs to be declared in order for it to be returned at the end of the function. And currently it will not be declared unless `check_specificity` is true-ish.

nimarafati replied:
Yes, you are right! I commented it while debugging and forgot to uncomment. I fix this.

nimarafati committed:
Set off_target_info to None when specificity check is off

pdmnyberg committed:
fix: include off target info in result from find_targets

nimarafati committed:
Update workflow and remove extraction steps

Removed old extraction steps and updated workflow to include run_plp_directrna.py with new parameters.
